src/uct/tcp/tcp_ep.c

```diff
@@ -50,11 +89,223 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_ep_t, uct_iface_t *tl_iface,
+                                uct_tcp_iface_t*, int,
+                                const struct sockaddr_in*)
+UCS_CLASS_DEFINE_NAMED_DELETE_FUNC(uct_tcp_ep_destroy, uct_tcp_ep_t, uct_ep_t)
+
+ucs_status_t uct_tcp_ep_create_connected(uct_iface_t *tl_iface,
+                                         const uct_device_addr_t *dev_addr,
+                                         const uct_iface_addr_t *iface_addr,
+                                         uct_ep_h *ep_p)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(tl_iface, uct_tcp_iface_t);
+    struct sockaddr_in dest_addr;
+    uct_tcp_ep_t *tcp_ep = NULL;
+    ucs_status_t status;
+
+    memset(&dest_addr, 0, sizeof(dest_addr));
+    dest_addr.sin_family = AF_INET;
+    dest_addr.sin_port   = *(in_port_t*)iface_addr;
+    dest_addr.sin_addr   = *(struct in_addr*)dev_addr;
+
+    status = uct_tcp_ep_create(iface, -1, &dest_addr, &tcp_ep);
+    if (status == UCS_OK) {
+        ucs_debug("tcp_ep %p: connected to %s:%d", tcp_ep,
+                  inet_ntoa(dest_addr.sin_addr), ntohs(dest_addr.sin_port));
+        *ep_p = &tcp_ep->super.super;
+    }
+    return status;
+}
+
+void uct_tcp_ep_mod_events(uct_tcp_ep_t *ep, uint32_t add, uint32_t remove)
+{
+    int new_events = (ep->events | add) & ~remove;
+
+    if (new_events != ep->events) {
+        ep->events = new_events;
+        ucs_trace("tcp_ep %p: set events to %c%c", ep,
+                  (new_events & EPOLLIN)  ? 'i' : '-',
+                  (new_events & EPOLLOUT) ? 'o' : '-');
+        uct_tcp_ep_epoll_ctl(ep, EPOLL_CTL_MOD);
+    }
+}
+
+static unsigned uct_tcp_ep_send(uct_tcp_ep_t *ep)
+{
+    size_t send_length;
+    ucs_status_t status;
+
+    send_length = ep->length - ep->offset;
+    ucs_assert(send_length > 0);
+
+    status = uct_tcp_send(ep->fd, ep->buf + ep->offset, &send_length);
+    if (status < 0) {
+        return 0;
+    }
+
+    ucs_trace_data("tcp_ep %p: sent %zu bytes", ep, send_length);
+
+    ep->offset += send_length;
+    if (ep->offset == ep->length) {
+        ep->offset = 0;
+        ep->length = 0;
+    }
+
+    return send_length > 0;
+}
+
+unsigned uct_tcp_ep_progress_tx(uct_tcp_ep_t *ep)
+{
+    uct_pending_req_priv_t *priv;
+    unsigned count = 0;
+
+    ucs_trace_func("ep=%p", ep);
+
+    if (ep->length > 0) {
+        count += uct_tcp_ep_send(ep);
+    }
+
+    uct_pending_queue_dispatch(priv, &ep->pending_q, uct_tcp_ep_can_send(ep));
+
+    if (uct_tcp_ep_can_send(ep) && ucs_queue_is_empty(&ep->pending_q)) {
+        uct_tcp_ep_mod_events(ep, 0, EPOLLOUT);
+    }
+
+    return count;
+}
+
+unsigned uct_tcp_ep_progress_rx(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr;
+    ucs_status_t status;
+    size_t recv_length;
+    ssize_t remainder;
+
+    ucs_trace_func("ep=%p", ep);
+
+    /* Receive next chunk of data */
```

Comment:
can it interfere with send?  Or i'm missing something

Reply:
endpoint structure is used either for send or receive (but not both), so no